### PR TITLE
Changes to allow Tilkynna runs on multi instance environment

### PR DESCRIPTION
**Pull request by melissapalmer**
**This PR includes:** 
- adding a processed_by column to the _reports.generated_report table. Which can be used to tell which instance processed the generate_report request. 
- correct setting the retry count correctly
- correct ability to config out the Scheduler poolSize
- correct check for correlationIdsExist 
- simplify the structure of query for getting PENDING items

### CODE REVIEW

File: src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java [hunk 1]
```diff
@@ -36,14 +37,37 @@
             "         FROM _reports.generated_report r " + //
             "         JOIN _reports.destination d ON r.destination_id = d.destination_id " + //
             "       WHERE cast(report_status AS varchar) = 'PENDING'  " + //
-            "       ORDER BY CASE d.downloadable WHEN true THEN 1 END, retry_count DESC, requested_at ASC LIMIT 30 FOR UPDATE SKIP LOCKED " + //
+            "       ORDER BY CASE d.downloadable WHEN true THEN 1 END, retry_count DESC, requested_at ASC LIMIT :batchSize FOR UPDATE SKIP LOCKED " + //
```
Comment (gm7) [inline src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java]: good to parameterize `batchSize`

File: src/main/java/org/tilkynna/common/utils/ApplicationInstance.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.net.UnknownHostException;
+
+import org.springframework.stereotype.Component;
+
+@Component
+public class ApplicationInstance {
+    private ApplicationInstance() {
+    }
+
+    public static String name() {
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/common/utils/ApplicationInstance.java]: The app instance name is re-calculated on each generation of a report.
Suggest to annotate @Provides for a bean like something like hostname.
Then, while we're here - can the name of the instance not come from config as well?
As in:
1. Check for config for instance name, use it
2. If instance name not in config, use InetAdd... getHostName() as above

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java [hunk 3]
```diff
@@ -93,6 +92,7 @@ public void pushGenerateReportToThreadPoolForProcessing(UUID correlationId) {
         }
 
         log.debug(String.format("start update status for ReportEntity correlationId [%s] on Thread [%s]", correlationId, Thread.currentThread().getName()));
+        generatedReportEntity.setProccesedBy(String.format("Instance [%s] on Thread [%s]", ApplicationInstance.name(), Thread.currentThread().getName()));
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: Here the an example of where ApplicationInstance.name is recalced

File: src/main/java/org/tilkynna/common/utils/ApplicationInstance.java (hunk 2 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/common/utils/ApplicationInstance.java]: It cannot come from config, as the instances are being spun up dynamically in Docker/K8s. 
Will make a change to stop the recalculation.

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java (hunk 3 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: ok see comments against above on this topic

File: src/main/java/org/tilkynna/common/utils/ApplicationInstance.java (hunk 2 above)
Comment (mvniekerk) [reply src/main/java/org/tilkynna/common/utils/ApplicationInstance.java]: Cool - as you've said a shared singleton bean will work

Comment (melissapalmer) [reply src/main/java/org/tilkynna/common/utils/ApplicationInstance.java]: This class is a Spring Bean (as @Component is used). Spring beans are singleton by default. 
https://docs.spring.io/spring-framework/docs/current/spring-framework-reference/core.html#beans-factory-scopes
I have changed in this commit https://github.com/GrindrodBank/tilkynna/compare/0bcd3b0aaafd...b2c50b087f76 to only call 'InetAddress.getLocalHost().getHostName();' on initialisation of the bean.

**Pull request merged**